Objects and Object Constructors: Add example of grouping functionality with objects (#27381)
TheOdinProject/curriculum: javascript/organizing_your_javascript_code/objects_and_object_constructors.md

// example two
const playerOne = {
  name: "tim",
  marker: "X"
};

const playerTwo = {
  name: "jenn",
  marker: "O"
};
```

> At first glance, the first doesn’t seem so bad.. and it actually takes fewer lines to write than the example using objects, but the benefits of the second approach are huge! Grouping related data together into objects allows you to pass the data around easily. Let me demonstrate:

Reviewer: ```suggestion
At first glance, the first doesn't seem so bad.. and it actually takes fewer lines to write than the example using objects, but the benefits of the second approach are huge! Grouping related data together into objects allows you to pass the data around easily. Let me demonstrate:
```
Somehow the apostrophes in this PR were adjusted to a different character. You'll want to fix all these to have them be how they were originally.

Author: This and the javascript/`js` code block changes were probably due to me copying/pasting from a markdown preview tool. I probably should have fixed the `js` code tag when I noticed it in the dif, but I did not. Didn't notice the apostrophes. Consider them fixed.

```js
const lightbulb = {
  lightColor: 'fluorescent white',  // this lightbulb is white,
  lit: false                        // and is currently 'off'
}
```

You may want to have the ability to switch a lightbulb from it's unlit state to it's lit state, or vice-versa. To do that, you might add a *method*.

The easiest way to get started creating methods to interact with your objects might be combining Object Literal syntax with JavaScript's `this` keyword. The `this` keyword is used to refer to the object a particular method is called from.

The following is an example of using the `this` keyword to add two methods to our object, `switchOn`, and `switchOff`:

```js
const lightbulb = {
  lightColor: 'fluorescent white',
  lit: false,

  // shorthand syntax for adding methods to objects
  switchOn() {
    // return false to indicate the light was already on
    if(this.lit === true) return false

    // return true if the state of the light changed to be on
    this.lit = true
    return true

>   },

Reviewer: This code would seem to violate the [command query separation](https://en.wikipedia.org/wiki/Command%E2%80%93query_separation) principle. Potentially not a big deal, just wondering why you think this should be returning a value AND setting the `lit` property?

Author: It's just a decision I made when writing the code. Haven't exactly heard of that principle previously.

I've personally just found having some commands return some kind of status to be useful. Interesting that you picked this one out, but not the RPS object. The usefulness of this isn't as clear, though, and you may be right that there's not a huge reason to include the additional code.

I do see what that principle is saying and what it's trying to avoid though- especially when the opposite is true and you want to perform some query without modification to the object.

    this.lit = false
    return true
  }
}

lightbulb.switchOn() // true - we switched it on
lightbulb.lit // true - the object has changed to reflect that!
```

These methods use the `this` keyword to refer to the object they get called from (`lightbulb`). The `this` keyword can be used to access and modify properties of an object in exactly the same way you would for any other variable that points to an object.

Feel free to copy this code in the console and experiment with it! If you're inclined, perhaps you could create a method to change the color of the light, as if it's one of those fancy RGB LEDs those gamer nerds and keyboard enthusiasts seem to like so much.

Moving past physical objects, we could also try to describe something like a game as an object. Since we've already explored Rock Paper Scissors in Foundations, let's use that as an example.

A rock paper scissors game might involve a couple basic things:

- Players' scores
- The ability to play a round (and playing a round should update a player's score)

And might also include a couple nice-to-haves

- The ability to determine the current winning player
- The ability to restart the game

> So, at it's most basic, an object that represents the game might look something like this (assuming we're playing against a comuter player):

Reviewer: ```suggestion
So, at it's most basic, an object that represents the game might look something like this (assuming we're playing against a computer player):
```

      (playerChoice === 'paper' && computerChoice === 'rock') ||
      (playerChoice === 'scissors' && computerChoice === 'paper')
    ) {
      return true
    } else {
      return false
    }
  },
  playerScore: 0,
  computerScore: 0,
  playRound(playerChoice) {
    // if an invalid choice is chosen, throw an error
    if(!this._options.includes(playerChoice)) {
      throw new Error(`Expected 'rock', 'paper', or 'scissors', but got ${playerChoice}`)
    }
    
    // get the computer's choice
    const computerChoice = this._getRandomChoice()

    // determine the winner, apply points if necessary, and return who won
    if(this._isTie(playerChoice, computerChoice)) {
      return "tie"
    } else if(this._isPlayerWinner(playerChoice, computerChoice)) {
      this.playerScore++
      return "player"
    } else {
      this.computerScore++
      return 'computer'
    }
  },
  getWinningPlayer() {
    if(this.playerScore === this.computerScore) {
      return 'tie'
    } else if (this.playerScore > this.computerScore) {
      return 'player'
    } else {
      return 'computer'
    }
  },
  reset() {
    this.playerScore = 0;
    this.computerScore = 0;
  }
}
```

> Another name for these might also be **private properties**/**private methods**, and even though object literal syntax doesn't provide a way to truly make them private, you will later learn about other methods of creating objects that *can*.

Reviewer: I kind of wonder how useful it is to go over this kind property/method privacy. Mostly just because this more of a "legacy" approach and in modern JavaScript, you'd accomplish this through `class` or factory functions.

I know you reference "you'll learn a better way later", but I think showing the learner this has the potential to raise a lot of questions.

Author: I wanted this lesson to introduce the idea of using objects to organize code in the simplest way possible, and to implant some idea of how one might craft an interface for an object.

Learners are likely to be exposed to this idea of using `_` to denote private properties as they read through content the curriculum offers anyway, and I feel there is value in covering the idea up-front.

Perhaps with the RPS object, an example of how private properties are helpful could be provided, where the `playerScore`/`computerScore` could be private properties, and a function is used to retrieve them instead. This could illustrate how *if those properties were really private*, the intended state of the object could be protected.

In my experience helping learners, many of them just don't grasp the usefulness of using objects for grouping functionality *at all*. My hope would be, by having an early preface to some of the basic ideas, it'll be in the back of learners minds as they continue through this section. 

I also feel covering the idea up-front may help learners understand the tradeoffs of using object literals/constructors/factories/classes as we cover each of them.

   1. Read the article [Understanding Prototypes and Inheritance in JavaScript](https://www.digitalocean.com/community/tutorials/understanding-prototypes-and-inheritance-in-javascript) from Digital Ocean. This is a good review of prototype inheritance and constructor functions, featuring some examples.
   1. To go a bit deeper into both the chain and inheritance, spend some time with [JavaScript.Info's article on Prototypal Inheritance](http://javascript.info/prototype-inheritance). As usual, doing the exercises at the end will help cement this knowledge in your mind. Don't skip them! Important note: This article makes heavy use of `__proto__` which is not generally recommended. The concepts here are what we're looking for at the moment. We will soon learn another method or two for setting the prototype.
1. You might have noticed us using the this keyword in object constructors and prototype methods in the examples above.
   1. [JavaScript Tutorial's article on the `this` keyword](https://www.javascripttutorial.net/javascript-this/) covers how this changes in various situations, including examples with constructor functions and the prototype.

> 1. Read the article [[[Prototype]] vs \_\_proto\_\_ vs .prototype in JavaScript](https://medium.com/@eamonocallaghan/prototype-vs-proto-vs-prototype-in-javascript-6758cadcbae8)

Reviewer 2: ```suggestion
1. Read the article [[[Prototype]] vs \_\_proto\_\_ vs .prototype in JavaScript](https://medium.com/@eamonocallaghan/prototype-vs-proto-vs-prototype-in-javascript-6758cadcbae8).
```
